src/transformers/trainer.py

```diff
@@ -395,8 +404,8 @@ def get_train_dataloader(self) -> DataLoader:
         )
 
     def _get_eval_sampler(self, eval_dataset: Dataset) -> Optional[torch.utils.data.sampler.Sampler]:
-        if isinstance(eval_dataset, torch.utils.data.IterableDataset):
-            return None
+        if not isinstance(eval_dataset, collections.abc.Sized):
+            raise ValueError("eval_dataset must implement __len__")
```

Comment:
This error message won't be understandable for the user if we end up here due to a `test_dataset`. I think the assert should be done at the beginning of evaluate (if the `eval_dataset` passed is not None) and predict, so the error comes to the user with the argument named the same way. We can have both since it won't harm, but I think the top ones are important.

Reply:
Indeed.
`evaluate` and `predict` make use of `get_eval_dataloader` and `get_test_dataloader`, which in turn both call `_get_eval_sampler`.
`get_eval_dataloader` and `get_test_dataloader` are also _public_ methods, a user could legitimately call them. They will also create an exception if a user calls them on a not-sized dataset, when calling `len()` on the dataset.
I added the same test there. Which means a sized dataset will be tested negative twice, while a not-sized dataset will be tested positive by the called method, either `evaluate`, `predict`